packages/astro/src/core/pipeline.ts

```diff
@@ -1,71 +1,28 @@
-import type { ComponentInstance, EndpointHandler, MiddlewareHandler } from '../@types/astro.js';
-import { callEndpoint, createAPIContext } from './endpoint/index.js';
+import type { ComponentInstance, EndpointHandler } from '../@types/astro.js';
+import { renderEndpoint } from '../runtime/server/endpoint.js';
+import { attachCookiesToResponse } from './cookies/response.js';
+import { createAPIContext } from './endpoint/index.js';
 import { callMiddleware } from './middleware/callMiddleware.js';
 import { renderPage } from './render/core.js';
 import { type Environment, type RenderContext } from './render/index.js';
 
-type PipelineHooks = {
-	before: PipelineHookFunction[];
-};
-
-export type PipelineHookFunction = (ctx: RenderContext, mod: ComponentInstance | undefined) => void;
-
 /**
  * This is the basic class of a pipeline.
  *
  * Check the {@link ./README.md|README} for more information about the pipeline.
  */
 export class Pipeline {
-	env: Environment;
-	#onRequest?: MiddlewareHandler;
-	#hooks: PipelineHooks = {
-		before: [],
-	};
-
-	/**
-	 * When creating a pipeline, an environment is mandatory.
-	 * The environment won't change for the whole lifetime of the pipeline.
-	 */
-	constructor(env: Environment) {
-		this.env = env;
-	}
-
-	setEnvironment() {}
-
-	/**
-	 * A middleware function that will be called before each request.
-	 */
-	setMiddlewareFunction(onRequest: MiddlewareHandler) {
-		this.#onRequest = onRequest;
-	}
-
-	/**
-	 * Removes the current middleware function. Subsequent requests won't trigger any middleware.
-	 */
-	unsetMiddlewareFunction() {
-		this.#onRequest = undefined;
-	}
-	/**
-	 * Returns the current environment
-	 */
-	getEnvironment(): Readonly<Environment> {
-		return this.env;
-	}
+	constructor(
+		readonly environment: Environment,
+		readonly locals: App.Locals,
+		readonly request: Request,
+		readonly pathname: string,
+		readonly renderContext: RenderContext,
+		readonly middleware = environment.middleware
```

Comment:
The presence of so many arguments can become very difficult to maintain in the future, even with types. Can we pass an object instead?